### PR TITLE
refactor callinfo to simplify ctor magic

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**

### CODE REVIEW

File: src/_pytest/runner.py [hunk 1]
```diff
@@ -190,35 +191,45 @@ def check_interactive_exception(call, report):
...
+            result = func()
         except:  # noqa
-            self.excinfo = ExceptionInfo()
-        self.stop = time()
+            excinfo = ExceptionInfo()
+            if reraise is not None and excinfo.errisinstance(reraise):
+                raise
+            result = None
+        stop = time()
+        return cls(start=start, stop=stop, when=when, result=result, excinfo=excinfo)
```
Comment (blueyed) [inline src/_pytest/runner.py]: Now it re-raises `KeyboardInterrupt`, but previously it was swallowed.

```diff
diff --git i/src/_pytest/runner.py w/src/_pytest/runner.py
index 44905fd7..c20d077a 100644
--- i/src/_pytest/runner.py
+++ w/src/_pytest/runner.py
@@ -194,7 +194,7 @@ def call_runtest_hook(item, when, **kwds):
     return CallInfo.from_call(
         lambda: ihook(item=item, **kwds),
         when=when,
-        reraise=KeyboardInterrupt if item.config.getvalue("usepdb") else (),
+        reraise=KeyboardInterrupt if not item.config.getvalue("usepdb") else (),
     )
```

But reraises should be an iterable always maybe?

Comment (RonnyPfannschmidt) [reply src/_pytest/runner.py]: argh

**Comment by blueyed** [comment]
CI failures appear to be unrelated.
https://github.com/pytest-dev/pytest/issues/4366#issuecomment-437631933
For MacOS py37 rebasing should fix it.

**Comment by RonnyPfannschmidt** [comment]
Will squash/rebase later today

@nicoddemus PS this is roughly what i want to enable for nodes AS well

**Comment by blueyed** [comment]
How is #4292 meant to look with this?

```diff
diff --git i/src/_pytest/runner.py w/src/_pytest/runner.py
index 0d6e3dc3..045eb72d 100644
--- i/src/_pytest/runner.py
+++ w/src/_pytest/runner.py
@@ -15,6 +15,7 @@
 from .reports import CollectReport
 from .reports import TestReport
 from _pytest._code.code import ExceptionInfo
+from _pytest.outcomes import Exit
 from _pytest.outcomes import skip
 from _pytest.outcomes import Skipped
 from _pytest.outcomes import TEST_OUTCOME
@@ -190,10 +191,13 @@ def check_interactive_exception(call, report):
 def call_runtest_hook(item, when, **kwds):
     hookname = "pytest_runtest_" + when
     ihook = getattr(item.ihook, hookname)
+    reraise = (Exit,)
+    if not item.config.getvalue("usepdb"):
+        reraise += (KeyboardInterrupt,)
     return CallInfo.from_call(
         lambda: ihook(item=item, **kwds),
         when=when,
-        reraise=KeyboardInterrupt if not item.config.getvalue("usepdb") else (),
+        reraise=reraise,
     )
```

?

Or would `Exit` be handled there still further down?

I think this tights Exit and KeyboardInterrupt just a bit more (as it currently stands)?!

**Comment by RonnyPfannschmidt** [comment]
@blueyed correct - the signaling should be layered differently

**Comment by blueyed** [comment]
How then?  (please be more verbose)

**Comment by RonnyPfannschmidt** [comment]
@blueyed my own investigation on that is not finished as the layering and the structure of calls, hooks and nodes is intertwined with this detail issue and its not clear how to layer and integrate them - i took a while to think if something comes to mind, but right no good solution seems to come to it

**Comment by blueyed** [comment]
But you would like to keep this PR as is?
Can you then please rebase and get it merged already, so that we can work from there?

(but IMHO given the above diff it does not really make it any better than https://github.com/pytest-dev/pytest/pull/4292/files#diff-4d6d922c3b97a9eae157f8f653627b13 would have done it)

**Comment by RonnyPfannschmidt** [comment]
pushed my squash/rebase - i might have invalidated the repr while in call test and should take a look at removing that again - will do more after i fetched the kid from daycare

**Comment by RonnyPfannschmidt** [comment]
yup, had to take it out again

File: testing/test_runner.py [hunk 2]
```diff
@@ -487,30 +487,20 @@ def test_report_extra_parameters(reporttype):
 
 
 def test_callinfo():
-    ci = runner.CallInfo(lambda: 0, "123")
+    ci = runner.CallInfo.from_call(lambda: 0, "123")
     assert ci.when == "123"
     assert ci.result == 0
     assert "result" in repr(ci)
-    assert repr(ci) == "<CallInfo when='123' result: 0>"
+    ci = runner.CallInfo.from_call(lambda: 0 / 0, "123")
```
Comment (blueyed) [inline testing/test_runner.py]: This test should not be changed.

File: testing/test_runner.py [hunk 3]
```diff
@@ -487,30 +487,20 @@ def test_report_extra_parameters(reporttype):
...
-def test_callinfo_repr_while_running():
-    def repr_while_running():
-        f = sys._getframe().f_back
-        assert "func" in f.f_locals
-        assert repr(f.f_locals["self"]) == "<CallInfo when='when' result: '<NOTSET>'>"
-
-    ci = runner.CallInfo(repr_while_running, "when")
-    assert repr(ci) == "<CallInfo when='when' result: None>"
-
-
```
Comment (blueyed) [inline testing/test_runner.py]: This test should be kept (in general, i.e. might need to be adjusted of course).

Comment (RonnyPfannschmidt) [reply testing/test_runner.py]: this test no longer makes sense, we no longer have callinfo objects that are invalid

File: testing/test_runner.py (hunk 2 above)
Comment (RonnyPfannschmidt) [reply testing/test_runner.py]: good catch

**Comment by RonnyPfannschmidt** [comment]
appveyor is codecov related

File: src/_pytest/runner.py [hunk 4]
```diff
@@ -189,43 +190,57 @@ def check_interactive_exception(call, report):
...
-            status = "exception: %s" % str(self.excinfo.value)
+        if self.excinfo is not None:
+            status = "exception"
+            value = self.excinfo.value
         else:
-            result = getattr(self, "result", "<NOTSET>")
-            status = "result: %r" % (result,)
-        return "<CallInfo when=%r %s>" % (self.when, status)
+            # TODO: investigate unification
+            value = repr(self._result)
```
Comment (nicoddemus) [inline src/_pytest/runner.py]: `safe_repr`?

Comment (RonnyPfannschmidt) [reply src/_pytest/runner.py]: wasn't used in the original

**Comment by nicoddemus** [comment]
@RonnyPfannschmidt this is conflicting with the recent merge for `ExceptionInfo`, could you please take a look when you get the chance?

**Comment by RonnyPfannschmidt** [comment]
will do so either in the late afternoon or tommorow

**Pull request merged**